MLIR#382: Remove hip dependency from mlir/lib

zhanglx13 opened the pull request:
1. All components in mlir/lib (except the rocm-runtime-wrappers) do not depend on BackendUtils, hence they do not depend on hip.
2. IsaNameParser is renamed as IsaNameSplitter and moved to Dialect/MIOpen/utility. 
4. IsaNameSplitter is compiled into MLIRMIOpenUtility.
5. mlir/include/mlir/ExecutionEngine is empty, so it is removed.
6. hipGetDeviceProperties() is invoked directly in mlir-rocm-runner

Any code that uses IsaNameSplitter needs to
1. Add #include "mlir/Dialect/MIOpen/utility/IsaNameSplitter.h", and
2. link to MLIRMIOpenUtility

zhanglx13 committed:
Remove unused Backendutils includes and lib dependencies

zhanglx13 committed:
Take hip-related functionality from Backendutils to mlir-rocm-runner

So that BackendUtils.h/cpp is ready to be removed.

zhanglx13 committed:
Remove BackendUtils.h and .cpp

zhanglx13 committed:
Move IsaNameParser to Dialect/MIOpen/Generator

IsaNameParser is moved to Dialect/MIOpen/Generator because
Conv2dGenerator.cpp is the only componet in /mlir/lib that uses
IsaNameParser.

Now IsaNameParser is compiled into a library named
LLVMIsaNameParser. Any code that uses IsaNameParser should:
1. Add #include "mlir/Dialect/MIOpen/Generator/IsaNameParser.h"
2. link to LLVMIsaNameParser

Since mlir/lib/ExecutionEngine and mlir/include/mlir/ExecutionEngine
are empty, they are also removed.

krzysz00 commented on mlir/lib/Conversion/GPUToMIGraphX/CMakeLists.txt:
```diff
@@ -19,5 +19,4 @@ target_link_libraries(MLIRGPUToMIGraphX
   MLIRMIGraphX
   MLIRGPUToGPURuntimeTransforms
   MLIRGPUToROCDLTransforms
-  LLVMROCmBackendUtils
```
Let's call it MLIRRocmBackendUtils?

zhanglx13 replied:
@jerryyin What do you think about the naming?

jerryyin replied:
I'm no pro to naming but `backend utils` seems too broad and vague compared to what it does. The only utility in this component has nothing to do with `backend`. I'd just get rid of `backend` in the naming.

jerryyin replied:
Just as FYI, we talked a second ago and sounds like a better suited name is: `IsaNameSplitter`. This is as close as it can get in what this component does.

krzysz00 commented on mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp:
```diff
@@ -1,1376 +0,0 @@
-//===- rocm-runtime-wrappers.cpp - MLIR ROCM runner wrapper library -------===//
```
... hold on, don't we still use these functions in tests? @sjw36 

I'm surprised this removal didn't break the build

zhanglx13 replied:
Now the tests are using mlir_rocm_runtime from upstream according to a recent commit PR631

krzysz00 replied:
Not all the functions the tests need are in that file though

zhanglx13 replied:
Thank you for pointing that out. I'll try to fix it.

krzysz00 replied:
The question will be ... do the tests actually run with this file deleted?

zhanglx13 replied:
No. It turns out I was testing without removing the rocm-runtime-wrapper library.
I brought it back and will submit again soon after the testing is done.

jerryyin replied:
Potentially I'd like to see us not having to maintain an additional "rocm-runtime-wrapper` (versus the upstream duplicate). I think what can be done (after this PR is merged) is:
 - file an upstream PR to add in whatever is necessary there
 - file another PR to get rid of `rocm-runtime-wrapper` duplicate in our side

krzysz00 replied:
I agree, but that's non-trivial in that `rocm-runtime-wrapper` here has the convolution validation code.

We'd need to do the thing where the convolution code gets moved into what I'm thinking of as the "mlir-miopen-test-lib.mlir" - a bunch of static MLIR functions that we can copy into tests. It's something I've been wanting to do for a while but haven't gotten around to.

(This also relates to @sjw36's plan to move more of the miopen-gen generated code to `gpu.*` ops ... which is vaguely in progress and I think requires a `gpu.free` op being added upstream)

jerryyin replied:
@krzysz00 Good point. I just forgot about the fact that validation related code are in here as well. 

So let me rephrase, in the upcoming refactor to decouple these two:
 - all `mgpu.*` function moved (added) to external repo
 - all `mcpu.*` function and its utils remain (and potentially, we can put this to a new runtime library `rocm-validation-wrapper`)

Just so that we are clear on the boundary of the two components.

krzysz00 reviewed (commented):
The movement looks reasonable, but I have one naming nitpick and I'm surprised that the removal of the runtime wrappers went through ...

zhanglx13 committed:
Revert "Remove BackendUtils.h and .cpp"

Only bring back rocm-runtime-wrappers library

This reverts commit eb77a354d922ed5e3b7943cae418861cd42463ed.

zhanglx13 committed:
Brought back the placeholder CMakeLists.txt

zhanglx13 committed:
Rename the IsaNameParser library as MLIRRocmBackendUtils

krzysz00 commented on mlir/lib/Dialect/MIOpen/Generator/CMakeLists.txt:
```diff
@@ -37,3 +41,7 @@ target_link_libraries(MLIRMIOpenConv2dGenerator
   ${LIBS}
 )
+
+add_llvm_library(MLIRRocmBackendUtils
```
add_llvm_library is absolutely not the function you want - that has the implication that the component is part of LLVM, not MLIR. `add_mlir_dialect_library`, maybe?

zhanglx13 replied:
According to the definition of mlir_add_dialect_library, this function declares a library associated with a dialect. Should we instead use the more general function mlir_add_library for IsaNamePaser since it is more like a utility function?

krzysz00 replied:
This is a library related to the `MIOpen` dialect, given which folder it's in

krzysz00 commented on mlir/tools/mlir-rocm-runner/mlir-rocm-runner.cpp:
```diff
@@ -33,9 +32,25 @@
+#include "mlir/Dialect/GPU/GPUDialect.h"
+#include "mlir/Dialect/GPU/Passes.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
```
Can we check to make sure that these includes are all actually needed?

zhanglx13 replied:
Actually the GPUDialect.h and Pass.h are not needed.

krzysz00 replied:
If you could drop them?

zhanglx13 replied:
Yes, I did.

krzysz00 commented on mlir/tools/mlir-rocm-runner/mlir-rocm-runner.cpp:
```diff
@@ -61,6 +76,25 @@ static cl::opt<bool>
+
+#if __INCLUDE_HIP__
+namespace {
+    void getGpuGCNArchName(hipDevice_t device, std::string &gcnArchName) {
```
`static` over anonymous namespaces for things like this, per upstream style we haven't always known about or followed well

And you'll want to `git clang-format` after staging and before committing (or `git clang-format miopen-dialect` after committing) or else your changes will get bounced back by CI for not being formatted correctly.

zhanglx13 replied:
Thank you @krzysz00 for the advice. I'll make sure I run it before submitting commits.

As per the discussion with @jerryyin, the next step will be to remove calling hipGetDeviceProperties() inside mlir-rocm-runner to avoid the out-of-handle problem when running multiple rocm instances concurrently. 
I'll check with the coding style to learn more about things like this.

krzysz00 replied:
There's no need to dive deeply into coding style (except for the bits that can't be auto-checked - there's an LLVM style guide we probably all collectively want to review, myself included) - `clang-format` is good at its job

krzysz00 reviewed (commented):
Overall looks reasonable and it's definitely an important cleanup, just a few things here and there

zhanglx13 committed:
Minor fix

1. Replaced add_llvm_library with add_mlir_library
2. in rocm-runner, removed ununsed headers
3. in rocm-runner, replaced anonymous namespace with static around getGpuGCNArchName

jerryyin commented on mlir/lib/Dialect/MIOpen/Generator/CMakeLists.txt:
```diff
@@ -37,3 +41,7 @@ target_link_libraries(MLIRMIOpenConv2dGenerator
+
+add_mlir_library(MLIRRocmBackendUtils
+  IsaNameParser.cpp
+)
```
Can we make the `IsaNameParser` to be a compilation unit of `Conv2dGenerator`?

jerryyin commented on mlir/lib/Dialect/MIOpen/Generator/CMakeLists.txt:
```diff
@@ -1,3 +1,7 @@
+set(LLVM_OPTIONAL_SOURCES
+  Conv2dGenerator.cpp
+  IsaNameParser.cpp
+)
```
I'd prefer to get rid of this (and not use `LLVM_*` in mlir CmakeLists)

krzysz00 replied:
`LLVM_OPTIONAL_SOURCES` is a pretty widely-used thing, IIRC

But I don't think we'll need id if we're using `add_mlir_dialect_library` calls

krzysz00 replied:
Overall, we could just add the library to the `Conv2DGenerator`. Or move it ta `utils/IsaNameParses.cpp` and add it to the MIOpen utils library (next to `loweringUtils`)

jerryyin commented on mlir/tools/mlir-rocm-runner/CMakeLists.txt:
```diff
@@ -4,14 +4,48 @@ set(LLVM_OPTIONAL_SOURCES
+    message(STATUS "ROCm HIP runtime lib: ${ROCM_RUNTIME_LIBRARY}")
+  endif()
+
+  # Set HIP compile-time flags.
+  add_definitions(-D__HIP_PLATFORM_HCC__)
+  add_definitions(-D__INCLUDE_HIP__=1)
```
We probably can get rid of these.

krzysz00 replied:
... we do need to define HCC when including HIP

jerryyin replied:
We still do now? `HCC` should be a deprecated old flag and we should be using `hip-clang` now. But who knows.. if we still need it then leave it as-is.

On the other hand, get rid of `__INCLUDE_HIP__` because that's defined by us and is no longer necessary.

krzysz00 replied:
We never use `hipe-clang` anywhere in the build.
So we have to specify we're targetting HCC with the flag

jerryyin replied:
Just for clarification, by default ROCM ship with `hip-clang`, not hcc, so I'm surprised we still need the "hcc" flag.

zhanglx13 replied:
In hip_runtime.h (rocm 5.0.1), it requires the definition of __HIP_PLATFORM_AMD/HCC__ or __HIP_PLATFORM_NVIDIA/NVCC__.  Later it does say that __HCC__ and __HIP_PLATFORM_HCC__ are deprecated.

jerryyin replied:
@zhanglx13 Sounds good, and thanks for looking into this. Please go ahead with `__HIP_PLATFORM_AMD__`.

jerryyin commented on mlir/tools/mlir-rocm-runner/mlir-rocm-runner.cpp:
```diff
@@ -61,6 +74,28 @@ static cl::opt<bool>
 
+static constexpr const char kTargetTriple[] = "amdgcn-amd-amdhsa";
+
+#if __INCLUDE_HIP__
```
ditto here, we can probably get rid of the macro together with (#if, #else...) here.

jerryyin commented on mlir/include/mlir/ExecutionEngine/ROCm/BackendUitls.h:
```diff
@@ -1,46 +0,0 @@
-//===- BackendUtils.h - MLIR to C++ option parsing ---------------===//
```
Just to make sure, the entire `ROCm` sub-directory is gotten rid of, correct?

krzysz00 replied:
Not anymore, I think

jerryyin reviewed (commented):
This cleanup makes all the entangled dependencies nicer and cleaner, thanks for your work.

krzysz00 commented on mlir/tools/mlir-rocm-runner/mlir-rocm-runner.cpp:
```diff
@@ -33,9 +32,23 @@
 #include <cstdlib>
 #include <mutex>
 
+#if __INCLUDE_HIP__
```
We can probably drop the `#if __INCLUDE_HIP_` in that this file won't be compiled if HIP isn't a dependency.

krzysz00 reviewed (commented):
Ok, so, we're down to bikeshedding about names. Which is important, but, still, fundamentally, good work!

zhanglx13 committed:
Rename IsaNameParser to IsaNameSplitter and move it to MIOpen/utility/

- Move IsaNameParser.cpp to /mlir/lib/Dialect/MIOpen/utility/
- Add IsaNameParser into MLIRMIOpenUtility library
- Replace the linking to MLIRRocmBackendUtils to MLIRMIOpenUtility
- Rename IsaNameParser to IsaNameSplitter
  String-relace all occurence of IsaNameParser to IsaNameSplitter

zhanglx13 committed:
Refactor according to clang-format

zhanglx13 committed:
Fixed the header guard in IsaNameSplitter.h

jerryyin reviewed (approved):
LGTM

krzysz00 reviewed (approved):
LGTM!

zhanglx13 commented:
@jerryyin @krzysz00 There is a connection reset error when building MIOpen. Should we rerun the check or there is something wrong on the server?

jerryyin commented:
@zhanglx13 That's a fairly typical MIOpen dependency (boost) failure, you can re-trigger it.

zhanglx13 commented:
@jerryyin @krzysz00 Now all tests have passed. Could one of you merge the pull request? Or should I do it?

jerryyin commented:
@zhanglx13 Feel free to do it if:
 - All tests are passing
 - At least 1 reviewer approved
 - No foreseeable breakage to nightly tests
